security/src/main/java/io/airlift/security/utils/CloseableGSSContext.java

```diff
@@ -0,0 +1,180 @@
+                    options.put("doNotPrompt", "true");
+                    if (log.isDebugEnabled()) {
+                        options.put("debug", "true");
+                    }
+
+                    String ticketCache = System.getenv("KRB5CCNAME");
+                    if (ticketCache != null) {
+                        options.put("ticketCache", ticketCache);
+                        options.put("useTicketCache", "true");
+                        options.put("renewTGT", "true");
+                    }
+
+                    return new AppConfigurationEntry[] {
+                            new AppConfigurationEntry(Krb5LoginModule.class.getName(), REQUIRED, options)
+                    };
+                }
+            });
+            String servicePrincipal = KerberosUtil.getServicePrincipal(serviceName, serviceUri.getHost());
+            Subject subject = KerberosUtil.getClientSubject(loginContext);
+            context = KerberosUtil.getGssContext(subject, servicePrincipal);
+        }
+        catch (LoginException | UnknownHostException e) {
+            throw new AuthenticationException(e);
+        }
+    }
+
+    public GSSContext getContext()
+    {
+        return context;
+    }
+
+    @Override
+    public void close()
+    {
+        try {
+            context.dispose();
+            loginContext.logout();
+        }
+        catch (GSSException | LoginException e) {
+            throw new AuthenticationException(e);
+        }
+    }
+
+    private static class KerberosUtil
+    {
+        private static final GSSManager gssManager = GSSManager.getInstance();
+
+        private static final Oid SPNEGO_OID;
+        private static final Oid KERBEROS_OID;
+
+        static {
+            try {
+                SPNEGO_OID = new Oid("1.3.6.1.5.5.2");
+                KERBEROS_OID = new Oid("1.2.840.113554.1.2.2");
+            }
+            catch (GSSException e) {
+                throw new AssertionError(e);
+            }
+        }
+
+        /**
+         * Create Kerberos principal for a given service and hostname. It converts
+         * hostname to lower case. If hostname is null or "0.0.0.0", it uses
+         * dynamically looked-up fqdn of the current host instead.
+         *
+         * @param serviceName Service for which you want to generate the principal.
+         * @param hostName Fully-qualified domain name.
+         * @return Converted Kerberos principal name.
+         * @throws UnknownHostException If no IP address for the local host could be found.
+         */
+        public static String getServicePrincipal(String serviceName, String hostName)
+                throws UnknownHostException
+        {
+            requireNonNull(serviceName, "serviceName is null");
+            requireNonNull(hostName, "hostName is null");
+            InetAddress address = InetAddress.getByName(hostName);
+            String fqdn;
+            if (address.getHostName().toLowerCase().startsWith("localhost")) {
```

Comment:
Why `startsWith`? This should be just the hostname, not the `toString()` version.